src/rgw/driver/rados/rgw_bucket.cc

```diff
@@ -1445,10 +1446,36 @@ int RGWBucketAdminOp::remove_bucket(rgw::sal::Driver* driver, RGWBucketAdminOpSt
   if (ret < 0)
     return ret;
 
+  // check if the bucket is owned by my zonegroup, if not, refuse to remove it
+  if (bucket->get_info().zonegroup != driver->get_zone()->get_zonegroup().get_id()) {
+    ldpp_dout(dpp, 0) << "ERROR: The bucket's zonegroup does not match. "
+                      << "Removal is not allowed. Please execute this operation "
+                      << "on the bucket's zonegroup: " << bucket->get_info().zonegroup << dendl;
+    return -ERR_PERMANENT_REDIRECT;
+  }
```

Comment:
shouldn't this be restricted only to non-master zonegroups?

Reply:
I guess to delete a bucket we need to make sure the delete operation chain has been started by the bucket's zonegroup (so it can delete the objects and make sure the bucket is empty first). Then if it wasn't the master zonegroup it can be redirected to the master in order to delete the metadata and other stuff.

Reply:
Basically, if we let the delete happen for a bucket in another zonegroup by the master zonegroup, it can leave orphan objects in the data pool from that bucket.